server/src/main/java/org/apache/druid/server/coordinator/duty/RoundRobinIterator.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.server.coordinator.duty;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.druid.error.InvalidInput;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * A round-robin iterator that is backed by an ordered list of candidates containing no duplicates. The iterator
+ * iterates endlessly over all the candidates. The caller must explicitly terminate it.
+ * The iterator has the following properties:
+ * <ul>
+ *   <li> Starts with an initial random cursor position in an ordered list of candidates. </li>
+ *   <li> Invoking {@code next()} on {@link #getIterator()} is guaranteed to result in a deterministic order
+ *   unless the set of candidates change when {@link #updateCandidates(Set)} is called. When the candidates change,
+ *   the cursor is reset to a random position in the new list of ordered candidates. </li>
+ *   <li> Guarantees that no duplicate candidates are returned in two consecutive {@code next()} iterations. </li>
+ * </ul>
+ *
+ * <p>
+ * This class is not thread-safe and must be used from a single thread.
+ */
+@NotThreadSafe
+public class RoundRobinIterator
```

Comment:
Some suggestions to make this class nicer:

1. The iteration aspect of this class seems vaguely similar to `RoundRobinServerSelector.CircularServerList`. Let's just pull out `CircularServerList` as a generic class `CircularList<T>` and put it in the package `org.apache.druid.collections`.
2. Make `CircularList` implement `Iterable`.
3. Ideally, this class would have no public methods right now, but since `RoundRobinServerSelector` is already using `peekNext()` and `advanceCursor()`, let's keep them for now. I can later take a look at `EligibleServerSelector` to see if we can't do without these methods.
4.  I am not entirely sold on requiring to reset the cursor to a random position (it would be nice to have a unit test that verifies the spread after say 1000 runs and compare when we reset to random and don't reset at all).
  - Either way, the `CircularList` should expose a method `resetCursor(int position)`. The randomness should not be baked into this class and should instead be provided by the caller (in this case, the kill duty).
  - Sorting and randomization should be mutually exclusive. If we are going to randomize, there is no point in sorting the set.
5. We can also add a method `equalsSet(Set<>)`.
6. The requirement to avoid duplicates should be a part of the caller, i.e. kill duty. (I recall having suggested it to be a part of the `RoundRobinIterator` but this seems cleaner now 😛 ).

After all of this, the kill duty would look as follows:


```java
private String prevDatasourceKilled;

// I don't like this variable name myself, if something better comes to mind, use it
private CircularList<String> datasourceCircularKillList;

public void run()
{
   ...
   ...
   // This goes where we identify the datasourcesToKill in the duty
   final Set<String> datasourcesToKill = ...;
   if (datasourceCircularKillList == null ||
         !datasourceCircularKillList.equalsSet(datasourcesToKill)) {
       datasourceCircularKillList = new CircularList<>(datasourcesToKill);
       
       // optional
       datasourceCircularKillList.resetCursor(randomPosition);
   }
   ...
   ...
   // This goes where we iterate over the circular list
   for (String datasource : datasourceCircularKillList) {
       if (datasource.equals(prevDatasourceKilled)) {
          continue;
       } else {
          prevDatasourceKilled = datasource;
       }
       
       // Do the rest of the kill processing
   }
   ...
   ...
}
```

Reply:
Good suggestion! I borrowed some motivation from `RoundRobinServerSelector` :P 

I like the idea and started refactoring and then realized it's going to bloat the changes in more than one place than what's necessary for the scope of this PR. I think it'll be best to do the refactoring-only change in a follow up. What do you think?

Reply:
I should imagine the only extra refactor required would be moving `CircularServerList` out of `RoundRobinServerSelector`. Are there other changes involved?

Reply:
The code reorganization will also entail retesting `RoundRobinServerSelector`. I currently don't have the bandwidth to do that.

Reply:
Oh, that should be fine. There are enough UTs and sims in place to ensure that everything continues to work as expected. You need not test it out separately.

If you still don't feel comfortable touching `RoundRobinServerSelector`, you may leave it for now. I can take a look at it later.

But we should still use the `CircularList` in this PR as discussed above instead of the current approach (given that you agree with the `CircularList` design).

Reply:
Okay, thanks for the suggestions. I've added `CircularList` which is currently used by the kill duty. I haven't refactored `RoundRobinServerSelector` in this PR, so a few differences:
1. `advanceCursor()` is not public, as it's not required by the kill duty
2.  I didn't add `resetCursor()` to keep things simple. Since the kill duty also tracks `prevDatasourceKilled`, this seems sufficient

`CircularList` is a sorted list. I've noted this in the Javadocs, but we can perhaps consider renaming the class to `CircularSortedList`. What do you think?